platform-sdk/docs/proposals/TSS-Library/TSS-Library.md

```diff
@@ -0,0 +1,860 @@
+      |_windows
+        |_lib.dll
+```
+It will unzip and load the correct library packaged inside the jar and access it using JNI calls.
+The API interface is through JNI. Once built, this should become a runtime dependency.
+
+#### Swirlds Native Support
+##### Overview
+Our implementation of the `swirlds-cryptography-pairings-api` will use native compiled libraries under the hood accessed with JNI.
+Given the non-GA state of Project Panama, this proposal does not consider it, but it has not been discarded for future change proposals.
+
+While dealing with native code using JNI, there are two possible ways of making the library accessible to Java code:
+1. Make it available in the classpath: The library must be previously installed in the executing environment
+   and reference its location in the classpath env variable.
+2. Distribute it as an application dependency. The caveat here is that:
+
+   a. The distributed library must be compiled in every possible architecture on which the application will be executed.
+
+   b. The library must be unpackaged from the jar as it cannot be accessed while compressed.
+
+Two key points we are considering in the development of this solution are:
+a) ensuring that it is compatible with the Java module system.
+b) Ensure that dependent software does not require the installation of any additional dependencies other than what is distributed in the jars.
+
+That is why the native-support library will help load the native library by implementing option 2.
+For this to work, the native support library must be included as a dependency in the module accessing the JNI wrapping code.
+The native libraries will be accessed assuming the following structure in the module declaring the dependency:
+
+```
+ native-library-client.jar
+ |_WEB-INF
+    |_arch64
+      |_macos
+        |_libhedera_bls_jni.dylib
+    |_amd64
+      |_macos
+        |_libhedera_bls_jni.dylib
+      |_linux
+        |_libhedera_bls_jni.so
+      |_windows
+        |_libhedera_bls_jni.dll
+    |_x86
+      |_linux
+        |_libhedera_bls_jni.so
+      |_windows
+        |_libhedera_bls_jni.dll
+  ...
+```
+
+lib folder is organized in subfolders by the platform identifier, as returned by `System.getProperty("os.arch")` and `System.getProperty("os.name")`
+
+##### Constraints
+This module will not depend on hedera-services artifacts, so it cannot include logging, metrics, configuration, or any other helper module from that repo.
+
+##### Public API
+###### `LibraryDescriptionEntry`
+**Description**: Given that the compilation of a native library produces files with different names under different OS and architectures, we need a way to assemble a catalog for all possible forms our library will take.
+
+A record of 3 elements that defines the name of the binary file of the library to load in a specific system architecture and OS.
+
+**Example**:
+```java
+    new LibraryDescriptionEntry("macos", "aarch64", "libhedera_bls_jni.dylib");
```

Comment:
We should have constants for common OS and ARCH as part of the public API. Next to that `String getOs()` / `String getArch()` would be nive as wrappers around System.getProperty.

Reply:
I agree. Adding that.